crates/matrix-sdk/src/widget/widget_settings.rs

```diff
@@ -0,0 +1,217 @@
+    pub static CLIENT_ID: &str = "$org.matrix.msc2873.client_id";
+    pub static DEVICE_ID: &str = "$org.matrix.msc2873.matrix_device_id";
+    pub static BASE_URL: &str = "$org.matrix.msc4039.matrix_base_url";
+}
+/// Settings of the widget.
+#[derive(Debug, Clone)]
+pub struct WidgetSettings {
+    /// Widget's unique identifier.
+    pub id: String,
+
+    /// Whether or not the widget should be initialized on load message
+    /// (`ContentLoad` message), or upon creation/attaching of the widget to
+    /// the SDK's state machine that drives the API.
+    pub init_after_content_load: bool,
+
+    /// This contains the url from the widget state event.
+    /// In this url placeholders can be used to pass information from the client
+    /// to the widget. Possible values are: `$matrix_widget_id`, `$parentUrl`...
+    ///
+    /// # Examples
+    ///
+    /// e.g `http://widget.domain?username=$userId`
+    /// will become: `http://widget.domain?username=@user_matrix_id:server.domain`.
+    pub raw_url: String,
+}
+
+impl WidgetSettings {
+    /// Create the actual Url that can be used to setup the WebView or IFrame
+    /// that contains the widget.
+    ///
+    /// # Arguments
+    ///
+    /// * `room` - A matrix room which is used to query the logged in username
+    /// * `props` - Properties from the client that can be used by a widget to
+    ///   adapt to the client. e.g. language, font-scale...
+
+    pub async fn generate_url(
+        &self,
+        room: &Room,
+        props: ClientProperties,
+    ) -> Result<Url, url::ParseError> {
+        let profile = room
+            .client()
+            .account()
+            .get_profile()
+            .await
+            .unwrap_or(get_profile::v3::Response::new(None, None));
+        let avatar_url = profile.avatar_url.map(|url| url.to_string()).unwrap_or("".to_owned());
+        let device_id = room.client().device_id().map(|d| d.to_string()).unwrap_or("".to_owned());
+
+        Url::parse(
+            &self
+                .raw_url
+                .as_str()
+                .replace(url_props::WIDGET_ID, &self.id)
+                .replace(url_props::AVATAR_URL, &avatar_url)
+                .replace(url_props::DEVICE_ID, &device_id)
+                .replace(url_props::DISPLAY_NAME, &profile.displayname.unwrap_or("".to_owned()))
+                .replace(url_props::BASE_URL, &encode(&room.client().homeserver().await.as_str()))
+                .replace(url_props::USER_ID, room.own_user_id().as_str())
+                .replace(url_props::ROOM_ID, &encode(room.room_id().as_str()))
+                .replace(url_props::LANGUAGE, props.language.as_str())
+                .replace(url_props::CLIENT_THEME, props.theme.as_str())
+                .replace(url_props::CLIENT_ID, &props.client_id),
+        )
+    }
+
+    /// `WidgetSettings` are usually created from a state event.
+    /// (currently unimplemented)
```

Comment:
It seems implemented to me :-)?

Reply:
This implementation creates a widget without using matrix room state events.
It is using a list of fn parameters to generate a WidgetSettings object.
It should work like this `Room.getWidgets() -> Vec<WidgetSettings>`